backend/src/main/java/gov/cdc/usds/simplereport/service/sms/SmsService.java

```diff
@@ -46,17 +49,46 @@ void init() throws NumberParseException {
 
   @AuthorizationConfiguration.RequirePermissionStartTestWithPatientLink
   @Transactional(noRollbackFor = {TwilioException.class, ApiException.class})
-  public String sendToPatientLink(UUID patientLinkId, String text) throws NumberParseException {
+  public List<SmsDeliveryResult> sendToPatientLink(UUID patientLinkId, String text) {
     PatientLink pl = pls.getRefreshedPatientLink(patientLinkId);
-    String messageId = sendToPerson(pl.getTestOrder().getPatient(), text);
-    tmsRepo.save(new TextMessageSent(pl, messageId));
-    return messageId;
+
+    List<SmsDeliveryResult> smsSendResults = sendToPerson(pl.getTestOrder().getPatient(), text);
+
+    smsSendResults.forEach(
+        smsDeliveryResult -> {
+          if (!smsDeliveryResult.getDeliverySuccess()) {
+            return;
+          }
+
+          tmsRepo.save(new TextMessageSent(pl, smsDeliveryResult.getMessageId()));
+        });
+
+    return smsSendResults;
   }
 
-  private String sendToPerson(Person p, String text) throws NumberParseException {
-    String msgId = sms.send(new PhoneNumber(formatNumber(p.getTelephone())), fromNumber, text);
-    LOG.debug("SMS send initiated {}", msgId);
-    return msgId;
+  private List<SmsDeliveryResult> sendToPerson(Person p, String text) {
+    List<SmsDeliveryResult> smsSendResults = new ArrayList<>();
+
+    p.getPhoneNumbers().stream()
+        .forEach(
```

Comment:
**question:** Since we're `.stream().forEach()`'ing here to add items to a List, might it make more sense to `map()` and then `.collect()` them directly into that stream?

Reply:
Indeed it would! I was initially grossed out at the side-effect of sending the text message inside a `map()`, but given that it returns a usable value makes me feel OK about it. See 7fb7c8d.